.github/workflows/ubuntu.yml

```diff
@@ -0,0 +1,71 @@
+name: Ubuntu
+
+on: push
+
+env:
+  DOCKER_REGISTRY: ghcr.io
+
+jobs:
+  # Build the Docker image for Ubuntu using different versions of GCC.
+  gcc:
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        version:
+          - os: jammy
+            gcc: 12
+          - os: noble
+            gcc: 14
+    steps:
+      - name: Checkout repository
+        uses: actions/checkout@v4
+      - name: Login to GitHub Container Registry
+        run: echo "${{ secrets.GITHUB_TOKEN }}" | docker login ${DOCKER_REGISTRY} -u ${{ github.actor }} --password-stdin
+      - name: Determine the Docker image name.
+        run: |
+          # Convert the repository name to lowercase as the organization name is
+          # uppercase, which is not permitted by the Docker registry.
+          DOCKER_REPOSITORY=${GITHUB_REPOSITORY,,}
+          echo "DOCKER_IMAGE=${DOCKER_REGISTRY}/${DOCKER_REPOSITORY}/${{ matrix.version.os }}:gcc${{ matrix.version.gcc }}" >> $GITHUB_ENV
+      - name: Build the Docker image
+        working-directory: docker/ubuntu
+        run: |
+          DOCKER_BUILDKIT=1 docker build . \
+          --target gcc \
+          --build-arg UBUNTU_VERSION=${{ matrix.version.os }} \
+          --build-arg GCC_VERSION=${{ matrix.version.gcc }} \
+          --tag ${{ env.DOCKER_IMAGE }}
+      - name: Push the Docker image
+        run: docker push ${{ env.DOCKER_IMAGE }}
+
+  # Build the Docker image for Ubuntu using different versions of Clang.
+  clang:
+    runs-on: ubuntu-latest
+    strategy:
+      matrix:
+        version:
+          - os: noble
+            clang: 18
+          - os: noble
+            clang: 19
```

Comment:
How about also adding images for clang 16 and 17 (also available on Noble)

I don't love that we need to support these old compilers, but since we do declare support the these it makes sense we should also have a build image.